snooty/n.py

```diff
@@ -0,0 +1,359 @@
+from typing import (
+    Any,
+    Dict,
+    Tuple,
+    List,
+    NamedTuple,
+    MutableSequence,
+    Sequence,
+    Optional,
+    Iterator,
+    Type,
+    TypeVar,
+    Union,
+    Generic,
+)
+from dataclasses import dataclass, asdict
+
+
+__all__ = (
+    "Node",
+    "InlineNode",
+    "Code",
+    "Parent",
+    "InlineParent",
+    "Heading",
+    "Section",
+    "Directive",
+    "DirectiveArgument",
+    "Paragraph",
+    "TargetIdentifier",
+    "Role",
+    "RefRole",
+    "Text",
+)
+
+SerializableType = Union[None, bool, str, int, float, Dict[str, Any], List[Any]]
+SerializedNode = Dict[str, SerializableType]
+_T = TypeVar("_T")
+
+
+@dataclass
+class Node:
+    __slots__ = ("span",)
+    type = "node"
+
+    span: Tuple[int]
+
+    def serialize(self) -> SerializedNode:
+        result = {k: v for k, v in asdict(self).items() if v is not None}
+        del result["span"]
+        result.update(type=self.type, position={"start": {"line": self.span[0]}})
+        return result
+
+    def get_text(self) -> str:
+        """Return pure textual content from a given AST node."""
+        return ""
+
+    @property
+    def start(self) -> Tuple[int]:
+        return self.span
+
+    def verify(self) -> None:
+        """Perform optional validations on this node."""
+        pass
+
+
+_N = TypeVar("_N", bound=Node)
+
+
+@dataclass
+class InlineNode(Node):
+    __slots__ = ()
+
+
+@dataclass
+class Code(Node):
+    __slots__ = ("lang", "copyable", "emphasize_lines", "value")
```

Comment:
They're a bit of a useful hack for CPython --- if you're using a *smart* Python implementation like PyPy (we're not, but I'm interested in evaluating it once it gets to Python 3.7-support), then __slots__ is both unnecessary and does nothing since the compiler already uses a more efficient data model than "here's a dict have fun"